Add response specs

bombaywalla opened the pull request:
Added support for generating response specs in addition to parameter specs.
With rudimentary tests.

lispyclouds commented:
Hey hey @bombaywalla long time! Thanks a lot for this, will take a proper look and get back asap.

lispyclouds commented on src/navi/core.clj:
```diff
@@ -72,7 +75,9 @@
 (defmulti spec
   (fn [^Schema schema]
-    (first (.getTypes schema))))
+    (if-let [x (first (.getTypes schema))]
```
can be written as
```clojure
(or (first (.getTypes schema)) "null")
```

bombaywalla replied:
Okay.  Seems fair.

lispyclouds commented on src/navi/core.clj:
```diff
@@ -158,8 +168,60 @@
+    nil
+    (into {}
+          (map (fn lhm-mapper [map-entry]
+                 (vector (key-fn (.getKey ^Map$Entry map-entry))
```
will there be a case wherein we need to recursively convert this map? this seems to handle only the top level keys.

bombaywalla replied:
Good question.  But I believe not.

You can always use this function again inside the `val-fn` if recursion is needed.

I was trying to refactor out common code.
If you'd like, I can replace some of the code in `path-item->data` and `routes-from` to use this function instead.  Let me know if you want me to do that.

I have added a note in the docstring to say that this only converts top-level keys and vals.

lispyclouds replied:
Right. Yes, lets pull out the common code while we are at it. :)

lispyclouds replied:
Also I tried out something on a hunch:
```clojure
(import '[java.util.LinkedHashMap])

(def lhm (java.util.LinkedHashMap/new))

(LinkedHashMap/.put lhm 1 "foo")
(LinkedHashMap/.put lhm 2 "bar")

(into {} lhm) ;; this is enough
```
We don't need the mapping to a seq of vectors I think.

bombaywalla replied:
Done.  Though I renamed the function.

bombaywalla replied:
Done.

lispyclouds commented on src/navi/core.clj:
```diff
@@ -158,8 +168,60 @@
+  So, convert a string to a Long if relevant.
+  If the string is \"default\" then return it as a keyword, otherwise pass through.
+  Arguably, all non-integer status codes should be converted to keywords."
+  [s]
```
is it possible that the user supplies an int via the yaml/json? would this regex match still work?

bombaywalla replied:
The [OpenAPI Spec](https://spec.openapis.org/oas/v3.1.0.html#patterned-fields-0) specifically states that the response code should be a string.  And, the `ApiResponse` class is declared to be a map from `String` to `ApiResponse`
```
public class ApiResponses extends LinkedHashMap<String, ApiResponse>
```
I tried to `.put` an int as the key and I got an Exception.

So, I'm presuming that we will never get an int here, and only get a String.

lispyclouds replied:
Makes sense. Thanks for verifying!

lispyclouds commented on src/navi/core.clj:
```diff
@@ -158,8 +168,60 @@
+  Arguably, all non-integer status codes should be converted to keywords."
+  [s]
+  (cond (re-matches #"\d{3}" s) (Long/parseLong s)
+        (= "default" s) (keyword s)
```
lets return a hardcoded `:default`.

bombaywalla replied:
Okay.

lispyclouds commented on src/navi/core.clj:
```diff
@@ -158,8 +168,60 @@
+(defn media-type->data
+  "Convert a Java Schema's MediaType to a spec that Reitit will accept."
+  [^MediaType mt]
+  (let [schema (.getSchema mt)]
```
can inline the let.

bombaywalla replied:
Okay.

lispyclouds commented on src/navi/core.clj:
```diff
@@ -158,8 +168,60 @@
+(defn handle-media-type-key
+  "If the media type is \"default\", then return it as a keyword, otherwise pass through."
+  [s]
+  (if (= "default" s)
```
lets return a hardcoded `:default` here too.

bombaywalla replied:
Okay.

lispyclouds commented on test/navi/core_test.clj:
```diff
@@ -185,13 +213,22 @@
+                                           (doto (MediaType.)
+                                             (.setSchema (ObjectSchema.)))))))
+          responses (doto (ApiResponses.)
+                      (.put "200" response))
```
is an int possible here?

bombaywalla replied:
If you mean instead of the string "200", then no, as I explained above.  I tried to `.put` and int and got an exception.

lispyclouds replied:
All good.

bombaywalla commented:
Thanks for the feedback.
I've pushed a new commit with the changes discussed.

lispyclouds commented:
You could also take out the response coercions from the unsupported section on the readme too in this PR :)

bombaywalla commented:
Inspired by your experimentations with LinkedHashMap, I found that since a LinkedHashMap also implements HashMap, Clojure "just works" with it without any interop needed.  So I re-wrote `linked-hash-ma->clj-map` to just use `reduce-kv` and since it is no longer interop-dependent, I renamed it to `update-kvs` and changed the parameter ordering to keep the map first, like `update-keys` and `update-vals`.

I also refactored the other code to use `update-kvs`.  And I moved the definition of `update-kvs` to earlier in the file.

I updated the README to  reflect current reality.

See what you think of the current state and let me know if it can be made better.

Thanks!

lispyclouds commented on src/navi/core.clj:
```diff
@@ -27,6 +31,18 @@
+  [m key-fn val-fn]
+  (if (nil? m)
+    nil
+    (reduce-kv (fn kv-mapper [m k v]
```
This is neat! I hammocked a bit more on your code and came up with this:
```clojure
(into {} (for [[k v] m] [(key-fn k) (val-fn v)])
```
feels simpler to me? lemme know what you think? `reduce-kv` is indeed more relevant but also feels a bit overkill? not sure, im open to thoughts!

lispyclouds replied:
Also (when m ...) can be used here instead of (if (nil? m) ...).

bombaywalla replied:
Thanks for the additional suggestion.

My personal feeling is that `reduce-kv` is the more appropriate function in this situation, and reflects the intent more.  It is also faster (approximately 80 uS vs 130 uS), most likely because of the intermediate sequence generated by the `for`.  My feeling is to leave it as is (using `reduce-kv`).  However, I am happy to make the change to use the `for` version if you'd prefer.  Let me know.

FWIW, here is the transcript of the benchmarking session.

```clojure
navi.core> (require '[criterium.core :as c])
nil
navi.core> (def x (into {} (for [i (range 1000)] [(str "k" i) i])))
#'navi.core/x
navi.core> (defn lispyclouds-update-kvs
             "Update a map using `key-fn` and `val-fn`.
  Sort of like composing `update-keys` and `update-vals`.
  Unlike `update-keys` or `update-vals`, preserve nils."
             [m key-fn val-fn]
             (if (nil? m)
               nil
               (into {} (for [[k v] m] [(key-fn k) (val-fn v)]))))
#'navi.core/lispyclouds-update-kvs
navi.core> (c/quick-bench (update-kvs x identity identity))
Evaluation count : 7560 in 6 samples of 1260 calls.
             Execution time mean : 78.108056 µs
    Execution time std-deviation : 930.911015 ns
   Execution time lower quantile : 77.313815 µs ( 2.5%)
   Execution time upper quantile : 79.620446 µs (97.5%)
                   Overhead used : 1.594769 ns

Found 1 outliers in 6 samples (16.6667 %)
	low-severe	 1 (16.6667 %)
 Variance from outliers : 13.8889 % Variance is moderately inflated by outliers
nil
navi.core> (c/quick-bench (update-kvs x identity identity))
Evaluation count : 6024 in 6 samples of 1004 calls.
             Execution time mean : 87.853485 µs
    Execution time std-deviation : 14.268741 µs
   Execution time lower quantile : 79.987076 µs ( 2.5%)
   Execution time upper quantile : 112.570970 µs (97.5%)
                   Overhead used : 1.594769 ns

Found 1 outliers in 6 samples (16.6667 %)
	low-severe	 1 (16.6667 %)
 Variance from outliers : 47.5319 % Variance is moderately inflated by outliers
nil
navi.core> (c/quick-bench (lispyclouds-update-kvs x identity identity))
Evaluation count : 4938 in 6 samples of 823 calls.
             Execution time mean : 129.911469 µs
    Execution time std-deviation : 8.714619 µs
   Execution time lower quantile : 120.260354 µs ( 2.5%)
   Execution time upper quantile : 141.261373 µs (97.5%)
                   Overhead used : 1.594769 ns
nil
navi.core> (c/quick-bench (lispyclouds-update-kvs x identity identity))
Evaluation count : 4824 in 6 samples of 804 calls.
             Execution time mean : 130.544097 µs
    Execution time std-deviation : 5.086537 µs
   Execution time lower quantile : 125.795358 µs ( 2.5%)
   Execution time upper quantile : 137.748456 µs (97.5%)
                   Overhead used : 1.594769 ns
nil
navi.core> 
```

bombaywalla replied:
Agree on using the `when`.

lispyclouds replied:
Great work on the benchmarking and agree on keeping reduce-kv. Will merge as soon as I get to a computer.

lispyclouds commented on src/navi/core.clj:
```diff
@@ -12,11 +12,15 @@
            [io.swagger.v3.parser OpenAPIV3Parser]
            [io.swagger.v3.parser.core.models ParseOptions]))
 
+(set! *warn-on-reflection* true)
```
can we move this into the `(comment)` maybe? its useful but as part of a lib it might not be nice to turn this on for everyone importing this.

bombaywalla replied:
Personally, I'd leave the `*warn-on-reflection*` in.  But I see why other users of the library might not like that.  I will move them to a comment (I think there is one already).

lispyclouds commented on test/navi/core_test.clj:
```diff
@@ -11,7 +11,10 @@
+           [io.swagger.v3.oas.models.parameters Parameter PathParameter HeaderParameter QueryParameter RequestBody]
+           [io.swagger.v3.oas.models.responses ApiResponses ApiResponse]))
+
+(set! *warn-on-reflection* true)
```
same for this.

bombaywalla replied:
Fixed.

lispyclouds commented:
Thanks a lot for the efforts and it looks good! Lemme know what you think of my extra change and post that we should be good to merge.
